app/controllers/actions/SubmittedDataRetrievalActionImpl.scala

```diff
@@ -0,0 +1,93 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.actions
+
+import cats.data.EitherT
+import cats.implicits._
+import controllers.handleApiResult
+import models.common.{BusinessId, JourneyContext, Mtditid, UserId}
+import models.database.UserAnswers
+import models.domain.ApiResultT
+import models.errors.HttpError
+import models.journeys.Journey
+import models.requests.OptionalDataRequest
+import play.api.libs.json.{Format, JsObject, Json}
+import play.api.mvc.{ActionTransformer, Request}
+import repositories.SessionRepositoryBase
+import services.SelfEmploymentServiceBase
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendHeaderCarrierProvider
+import utils.Logging
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+trait SubmittedDataRetrievalAction extends ActionTransformer[OptionalDataRequest, OptionalDataRequest]
+
+class SubmittedDataRetrievalActionImpl[SubsetOfAnswers: Format](journeyContext: Mtditid => JourneyContext,
+                                                                selfEmploymentService: SelfEmploymentServiceBase,
+                                                                sessionRepository: SessionRepositoryBase)(implicit ec: ExecutionContext)
+    extends SubmittedDataRetrievalAction
+    with FrontendHeaderCarrierProvider
+    with Logging {
+
+  protected def executionContext: ExecutionContext = ec
+
+  protected[actions] def transform[A](request: OptionalDataRequest[A]): Future[OptionalDataRequest[A]] = {
+    val ctx: JourneyContext     = journeyContext(request.mtditid)
+    val hasAtLeastOneUserAnswer = hasAtLeastOneUserAnswerForJourney(ctx.businessId, ctx.journey, request.userAnswers)
+
+    if (hasAtLeastOneUserAnswer) {
+      Future.successful(request)
+    } else {
+      val existingUserAnswers = request.userAnswers.getOrElse(UserAnswers.empty(UserId(request.userId)))
+      upsertJourneyAnswers(ctx, request.request, existingUserAnswers).map { updated =>
+        request.copy(userAnswers = if (updated.isEmpty) None else Some(updated))
+      }
+    }
+  }
+
+  private def upsertJourneyAnswers[A](ctx: JourneyContext, request: Request[A], existingAnswers: UserAnswers): Future[UserAnswers] = {
+    val result: ApiResultT[UserAnswers] = for {
+      maybeJourneyAnswers <- selfEmploymentService.getSubmittedAnswers[SubsetOfAnswers](ctx)(implicitly[Format[SubsetOfAnswers]], hc(request))
+      updatedAnswers <- maybeJourneyAnswers.fold {
+        EitherT.pure[Future, HttpError](existingAnswers)
```

Comment:
Also, if are able to take out the Option like my comment above says, it will simplify this a little we won't need to accommodate for None case. What do you think? I may be missing something.

Reply:
None is when we don't have submitted values (request OK, but just no answers)